src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,126 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int INITIAL_CAPACITY = 16;
+    private static final int GROWTH_FACTOR = 2;
+    private static final float LOAD_FACTOR = 0.75f;
+    private int capacity;
+    private int threshold;
+    private Node<K, V>[] buckets;
+    private int size;
+
+    public MyHashMap() {
+        this.buckets = new Node[INITIAL_CAPACITY];
+        this.capacity = INITIAL_CAPACITY;
+        this.threshold = (int) (INITIAL_CAPACITY * LOAD_FACTOR);
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (key == null) {
+            putForNullKey(value);
+            return;
+        }
+        if (size >= threshold) {
+            resize();
+        }
+        int index = getIndex(key);
+        Node<K, V> bucket = buckets[index];
+        while (bucket != null) {
+            if (keyEquals(bucket.key, key)) {
+                bucket.value = value;
+                return;
+            }
+            bucket = bucket.next;
+        }
+        addNode(key, value, index);
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            return getValueForNullKey();
+        }
```

Comment:
I think that it's redundant. Because in  methods getIndex() and keyEquals(), we check null keys as well

Reply:
> I think that it's redundant. Because in methods getIndex() and keyEquals(), we check null keys as well

The solution you proposed would be more compact, but, I think for better performance, we need to use a special case for key==null. Because we will not need to allocate memory for the methods getIndex() and keyEquals().

Reply:
Okay, I agree with you)